File: chatuikit/src/main/java/com/cometchat/chatuikit/shared/views/searchbox/CometChatSearchBox.java
package com.cometchat.chatuikit.shared.views.searchbox;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.inputmethod.EditorInfo;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.StringDef;
import androidx.annotation.StyleRes;

import com.cometchat.chatuikit.databinding.CometchatSearchBoxBinding;
import com.cometchat.chatuikit.shared.interfaces.OnClick;
import com.cometchat.chatuikit.shared.resources.utils.Utils;
import com.google.android.material.card.MaterialCardView;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class CometChatSearchBox extends MaterialCardView {
    private static final String TAG = CometChatSearchBox.class.getSimpleName();
    private CometchatSearchBoxBinding binding;
    private @StyleRes int searchInputTextAppearance;
    private @ColorInt int searchInputTextColor;
    private @StyleRes int searchInputPlaceHolderTextAppearance;
    private @ColorInt int searchInputPlaceHolderTextColor;
    private Drawable searchInputStartIcon;
    private Drawable searchInputEndIcon;
    private @ColorInt int searchInputStartIconTint;
    private @ColorInt int searchInputEndIconTint;
    private OnSearch eventListener;
    private OnClick onSearchClick;

    public CometChatSearchBox(@NonNull Context context) {
        this(context, null);
    }

    public CometChatSearchBox(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CometChatSearchBox(@NonNull Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        if (!isInEditMode()) {
            inflateAndInitializeView();
        }
    }

    private void inflateAndInitializeView() {
        // Inflate the layout for this view
        binding = CometchatSearchBoxBinding.inflate(LayoutInflater.from(getContext()), this, true);

        // Reset the card view to default values
        Utils.initMaterialCard(this);

        binding.etSearch.setOnEditorActionListener((textView, i, keyEvent) -> {
            if (i == EditorInfo.IME_ACTION_SEARCH) {
                if (!textView.getText().toString().isEmpty()) {
                    if (eventListener != null) {
                        eventListener.onSearch(SearchState.Filter, textView.getText().toString());
                    }
                }
                return true;
            }
            return false;
        });

        binding.etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (eventListener != null) {
                    if (s.toString().isEmpty()) {
                        setSearchInputEndIconVisibility(GONE);
                    } else {
                        setSearchInputEndIconVisibility(VISIBLE);
                    }
                    eventListener.onSearch(SearchState.TextChange, s.toString());
                }
            }
        });

        binding.ivClear.setOnClickListener(view -> {
            binding.etSearch.setText("");
            if (eventListener != null) {
                eventListener.onSearch(SearchState.Clear, "");
            }
        });

        binding.searchBoxParent.setOnClickListener( view -> {
            if (onSearchClick != null) {
                onSearchClick.onClick();
            }
        });

        binding.etSearch.setOnClickListener( view -> {
            if (onSearchClick != null) {
                onSearchClick.onClick();
            }
        });
    }

    public void setOnSearchClick(OnClick onSearchClick) {
        this.onSearchClick = onSearchClick;
    }

    public void setSearchInputEndIconVisibility(int visibility) {
        binding.ivClear.setVisibility(visibility);
    }

    public void setSearchInputText(String text) {
        binding.etSearch.setText(text);
    }

    public @StyleRes int getSearchInputTextAppearance() {
        return searchInputTextAppearance;
    }

    public void setSearchInputTextAppearance(@StyleRes int searchInputTextAppearance) {
        this.searchInputTextAppearance = searchInputTextAppearance;
        binding.etSearch.setTextAppearance(searchInputTextAppearance);
    }

    public @ColorInt int getSearchInputTextColor() {
        return searchInputTextColor;
    }

    public void setSearchInputTextColor(@ColorInt int searchInputTextColor) {
        this.searchInputTextColor = searchInputTextColor;
        binding.etSearch.setTextColor(searchInputTextColor);
    }

    public @StyleRes int getSearchInputPlaceHolderTextAppearance() {
        return searchInputPlaceHolderTextAppearance;
    }

    public void setSearchInputPlaceHolderTextAppearance(@StyleRes int searchInputPlaceHolderTextAppearance) {
        this.searchInputPlaceHolderTextAppearance = searchInputPlaceHolderTextAppearance;
    }

    public @ColorInt int getSearchInputPlaceHolderTextColor() {
        return searchInputPlaceHolderTextColor;
    }

    public void setSearchInputPlaceHolderTextColor(@ColorInt int searchInputPlaceHolderTextColor) {
        this.searchInputPlaceHolderTextColor = searchInputPlaceHolderTextColor;
        binding.etSearch.setHintTextColor(searchInputPlaceHolderTextColor);
    }

    public Drawable getSearchInputStartIcon() {
        return searchInputStartIcon;
    }

    public void setSearchInputStartIcon(Drawable searchInputStartIcon) {
        this.searchInputStartIcon = searchInputStartIcon;
        binding.ivSearch.setBackground(searchInputStartIcon);
    }

    public @ColorInt int getSearchInputStartIconTint() {
        return searchInputStartIconTint;
    }

    public void setSearchInputStartIconTint(@ColorInt int searchInputStartIconTint) {
        this.searchInputStartIconTint = searchInputStartIconTint;
        binding.ivSearch.setBackgroundTintList(ColorStateList.valueOf(searchInputStartIconTint));
    }

    public Drawable getSearchInputEndIcon() {
        return searchInputEndIcon;
    }

    public void setSearchInputEndIcon(Drawable searchInputEndIcon) {
        this.searchInputEndIcon = searchInputEndIcon;
        binding.ivClear.setBackground(searchInputEndIcon);
    }

    public @ColorInt int getSearchInputEndIconTint() {
        return searchInputEndIconTint;
    }

    public void setSearchInputEndIconTint(@ColorInt int searchInputEndIconTint) {
        this.searchInputEndIconTint = searchInputEndIconTint;
        binding.ivClear.setBackgroundTintList(ColorStateList.valueOf(searchInputEndIconTint));
    }

    /**
     * Sets the placeholder text for the search input field.
     *
     * @param placeholder The text to be set as the placeholder.
     */
    public void setSearchPlaceholderText(String placeholder) {
        if (placeholder != null) {
            binding.etSearch.setHint(placeholder);
        }
    }

    /**
     * Sets an event listener for search events.
     *
     * @param eventListener The event listener to be set.
     */
    public void addOnSearchListener(OnSearch eventListener) {
        if (eventListener != null) {
            this.eventListener = eventListener;
        }
    }

    public CometchatSearchBoxBinding getBinding() {
        return binding;
    }

    @Retention(RetentionPolicy.SOURCE)
    @StringDef({SearchState.Filter, SearchState.Clear, SearchState.TextChange})
    public @interface SearchState {
        String Filter = "filter";
        String Clear = "clear";
        String TextChange = "textchange";
    }

    public interface OnSearch {
        void onSearch(@SearchState String state, String text);
    }
}

File: chatuikit-kotlin/src/main/java/com/cometchat/uikit/kotlin/presentation/shared/searchbox/CometChatSearchBox.kt
package com.cometchat.uikit.kotlin.presentation.shared.searchbox

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.ImageView
import androidx.annotation.ColorInt
import androidx.annotation.StyleRes
import com.cometchat.uikit.kotlin.R
import com.cometchat.uikit.kotlin.databinding.CometchatSearchBoxBinding
import com.cometchat.uikit.kotlin.shared.resources.utils.Utils
import com.google.android.material.card.MaterialCardView

/**
 * CometChatSearchBox is a custom search input component with search icon,
 * clear button, and customizable styling.
 */
class CometChatSearchBox @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : MaterialCardView(context, attrs, defStyleAttr) {

    companion object {
        private val TAG = CometChatSearchBox::class.java.simpleName
    }

    val binding: CometchatSearchBoxBinding

    @StyleRes private var searchInputTextAppearance: Int = 0
    @ColorInt private var searchInputTextColor: Int = 0
    @StyleRes private var searchInputPlaceHolderTextAppearance: Int = 0
    @ColorInt private var searchInputPlaceHolderTextColor: Int = 0
    private var searchInputStartIcon: Drawable? = null
    private var searchInputEndIcon: Drawable? = null
    @ColorInt private var searchInputStartIconTint: Int = 0
    @ColorInt private var searchInputEndIconTint: Int = 0

    private var eventListener: OnSearch? = null
    private var onSearchClick: (() -> Unit)? = null

    init {
        binding = CometchatSearchBoxBinding.inflate(LayoutInflater.from(context), this, true)
        Utils.initMaterialCard(this)
        setupListeners()
    }

    private fun setupListeners() {
        binding.etSearch.setOnEditorActionListener { textView, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                if (textView.text.toString().isNotEmpty()) {
                    eventListener?.onSearch(SearchState.FILTER, textView.text.toString())
                }
                true
            } else {
                false
            }
        }

        binding.etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                eventListener?.let { listener ->
                    if (s.toString().isEmpty()) {
                        setSearchInputEndIconVisibility(View.GONE)
                    } else {
                        setSearchInputEndIconVisibility(View.VISIBLE)
                    }
                    listener.onSearch(SearchState.TEXT_CHANGE, s.toString())
                }
            }
        })

        binding.ivClear.setOnClickListener {
            binding.etSearch.setText("")
            eventListener?.onSearch(SearchState.CLEAR, "")
        }

        binding.searchBoxParent.setOnClickListener {
            onSearchClick?.invoke()
        }

        binding.etSearch.setOnClickListener {
            onSearchClick?.invoke()
        }
    }

    fun setOnSearchClick(onClick: (() -> Unit)?) {
        this.onSearchClick = onClick
    }

    fun setSearchInputEndIconVisibility(visibility: Int) {
        binding.ivClear.visibility = visibility
    }

    fun setSearchInputText(text: String) {
        binding.etSearch.setText(text)
    }

    fun getSearchInputTextAppearance(): Int = searchInputTextAppearance

    fun setSearchInputTextAppearance(@StyleRes appearance: Int) {
        this.searchInputTextAppearance = appearance
        binding.etSearch.setTextAppearance(appearance)
    }

    fun getSearchInputTextColor(): Int = searchInputTextColor

    fun setSearchInputTextColor(@ColorInt color: Int) {
        this.searchInputTextColor = color
        binding.etSearch.setTextColor(color)
    }

    fun getSearchInputPlaceHolderTextAppearance(): Int = searchInputPlaceHolderTextAppearance

    fun setSearchInputPlaceHolderTextAppearance(@StyleRes appearance: Int) {
        this.searchInputPlaceHolderTextAppearance = appearance
        if (appearance != 0) {
            // Note: TextAppearance for hint is typically set via the same setTextAppearance
            // as it affects the overall text styling. The hint color is set separately.
            binding.etSearch.setTextAppearance(appearance)
        }
    }

    fun getSearchInputPlaceHolderTextColor(): Int = searchInputPlaceHolderTextColor

    fun setSearchInputPlaceHolderTextColor(@ColorInt color: Int) {
        this.searchInputPlaceHolderTextColor = color
        binding.etSearch.setHintTextColor(color)
    }

    fun getSearchInputStartIcon(): Drawable? = searchInputStartIcon

    fun setSearchInputStartIcon(icon: Drawable?) {
        this.searchInputStartIcon = icon
        binding.ivSearch.background = icon
    }

    fun getSearchInputStartIconTint(): Int = searchInputStartIconTint

    fun setSearchInputStartIconTint(@ColorInt tint: Int) {
        this.searchInputStartIconTint = tint
        binding.ivSearch.backgroundTintList = ColorStateList.valueOf(tint)
    }

    fun getSearchInputEndIcon(): Drawable? = searchInputEndIcon

    fun setSearchInputEndIcon(icon: Drawable?) {
        this.searchInputEndIcon = icon
        binding.ivClear.background = icon
    }

    fun getSearchInputEndIconTint(): Int = searchInputEndIconTint

    fun setSearchInputEndIconTint(@ColorInt tint: Int) {
        this.searchInputEndIconTint = tint
        binding.ivClear.backgroundTintList = ColorStateList.valueOf(tint)
    }

    /**
     * Sets the placeholder text for the search input field.
     */
    fun setSearchPlaceholderText(placeholder: String?) {
        placeholder?.let { binding.etSearch.hint = it }
    }

    /**
     * Sets an event listener for search events.
     */
    fun addOnSearchListener(listener: OnSearch?) {
        listener?.let { this.eventListener = it }
    }

    /**
     * Search state constants.
     */
    object SearchState {
        const val FILTER = "filter"
        const val CLEAR = "clear"
        const val TEXT_CHANGE = "textchange"
    }

    /**
     * Interface for search event callbacks.
     */
    fun interface OnSearch {
        fun onSearch(state: String, text: String)
    }
}
